libraries/typings-generator/src/TypingsGenerator.ts

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import {
+  FileSystem,
+  Terminal,
+  ConsoleTerminalProvider,
+  Path,
+  NewlineKind
+} from '@microsoft/node-core-library';
+import * as glob from 'glob';
+import * as path from 'path';
+import { EOL } from 'os';
+import * as chokidar from 'chokidar';
+
+/**
+ * @public
+ */
+export interface ITypingsGeneratorOptions<TTypingsResult = string> {
+  srcFolder: string;
+  generatedTsFolder: string;
+  fileExtensions: string[];
+  parseAndGenerateTypings: (fileContents: string, filePath: string) => TTypingsResult;
+  terminal?: Terminal;
+  filesToIgnore?: string[];
+}
+
+/**
+ * This is a simple tool that generates .d.ts files for non-TS files.
+ *
+ * @public
+ */
+export class TypingsGenerator {
+  protected _options: ITypingsGeneratorOptions;
+
+  public constructor(options: ITypingsGeneratorOptions) {
+    this._options = {
+      ...options
+    };
+
+    if (!this._options.generatedTsFolder) {
+      throw new Error('generatedTsFolder must be provided');
+    }
+
+    if (!this._options.srcFolder) {
+      throw new Error('srcFolder must be provided');
+    }
+
+    if (Path.isUnder(this._options.srcFolder, this._options.generatedTsFolder)) {
+      throw new Error('srcFolder must not be under generatedTsFolder');
+    }
+
+    if (Path.isUnder(this._options.generatedTsFolder, this._options.srcFolder)) {
+      throw new Error('generatedTsFolder must not be under srcFolder');
+    }
+
+    if (!this._options.fileExtensions || this._options.fileExtensions.length === 0) {
+      throw new Error('At least one file extension must be provided.');
+    }
+
+    if (!this._options.filesToIgnore) {
+      this._options.filesToIgnore = [];
+    }
+
+    if (!this._options.terminal) {
+      this._options.terminal = new Terminal(new ConsoleTerminalProvider({ verboseEnabled: true }));
+    }
+
+    this._options.fileExtensions = this._normalizeFileExtensions(this._options.fileExtensions);
+  }
+
+  public generateTypings(): void {
+    FileSystem.ensureEmptyFolder(this._options.generatedTsFolder);
```

Comment:
this can be problematic if someone sets the generatedTsFolder to be `src` or  another directory that shouldn't be cleaned, either accidentally or on purpose.

Reply:
It's mentioned in the documentation. I think this is reasonable given that it's documented.